crates/circuit/src/packed_instruction.rs

```diff
@@ -39,132 +42,375 @@ enum PackedOperationType {
+/// in exactly their 3 lowest bits.
+///
+/// # Safety
+/// Use [BitField::discriminant] to determine which field is valid for access.
+#[repr(C)]
+union BitField {
+    gate: StandardGateBits,
+    instruction: StandardInstructionBits,
+    pointer: PointerBits,
+}
+
+impl BitField {
+    // For good measure, we assert that the discriminant occupies exactly the 3 lowest bits on all
+    // of our BitField union members, using the handy constants generated by the bitfield-struct
+    // crate :)
+    const DISCRIMINANT_MASK: u64 = 0b111;
+    #[allow(clippy::assertions_on_constants)]
+    const _CHECK: () = {
+        assert!(
+            ((1 << StandardGateBits::DISCRIMINANT_BITS) - 1)
+                << StandardGateBits::DISCRIMINANT_OFFSET
+                == Self::DISCRIMINANT_MASK,
+            "(StandardGateBits) discriminant MUST be the 3 lowest bits!"
+        );
+        assert!(
+            ((1 << StandardInstructionBits::DISCRIMINANT_BITS) - 1)
+                << StandardInstructionBits::DISCRIMINANT_OFFSET
+                == Self::DISCRIMINANT_MASK,
+            "(StandardInstructionBits) discriminant MUST be the 3 lowest bits!"
+        );
+        assert!(
+            ((1 << PointerBits::DISCRIMINANT_BITS) - 1) << PointerBits::DISCRIMINANT_OFFSET
+                == Self::DISCRIMINANT_MASK,
+            "(PointerBits) discriminant MUST be the 3 lowest bits!"
+        );
+    };
 
-impl PackedOperation {
-    /// The bits representing the `PackedOperationType` discriminant.  This can be used to mask out
-    /// the discriminant, and defines the rest of the bit shifting.
-    const DISCRIMINANT_MASK: usize = 0b11;
-    /// The number of bits used to store the discriminant metadata.
-    const DISCRIMINANT_BITS: u32 = Self::DISCRIMINANT_MASK.count_ones();
-    /// A bitmask that masks out only the standard gate information.  This should always have the
-    /// same effect as `POINTER_MASK` because the high bits should be 0 for a `StandardGate`, but
-    /// this is defensive against us adding further metadata on `StandardGate` later.  After
-    /// masking, the resulting integer still needs shifting downwards to retrieve the standard gate.
-    const STANDARD_GATE_MASK: usize = (u8::MAX as usize) << Self::DISCRIMINANT_BITS;
-    /// A bitmask that retrieves the stored pointer directly.  The discriminant is stored in the
-    /// low pointer bits that are guaranteed to be 0 by alignment, so no shifting is required.
-    const POINTER_MASK: usize = usize::MAX ^ Self::DISCRIMINANT_MASK;
-
-    /// Extract the discriminant of the operation.
     #[inline]
     fn discriminant(&self) -> PackedOperationType {
-        ::bytemuck::checked::cast((self.0 & Self::DISCRIMINANT_MASK) as u8)
+        // SAFETY: we read (just!) the discriminant from any of the union's members,
+        // since we guarantee it is found in the same place for all bitfields.
+        unsafe { self.gate.discriminant() }
```

Comment:
The `StandardGateBits` struct is actually just a transparent tuple struct around a `u64`. And that `u64` is zero-initialized, so we should be safe here.

Reply:
I can buy that it isn't UB, though I don't think it's because of 0 initialisation (the u64 of `self` here isn't 0 - there's no `new` to be called, since we're interpreting a union field). If it's not UB, it's because the `StandardGate` type isn't constructed by the (internally generated) mask and shift of the getter, but the definition of the `struct` certainly makes it _look_ like you should think about the value as having been constructed.

I think I was wrong initially and it's not UB, but I'm still concerned about how non-obvious it is without having really thought about how `bitfield_struct` must work internally, which is more mental load.